packages/sdk/src/fetcher/core/index.ts

```diff
@@ -209,6 +219,133 @@ class Fetcher implements ICoreFetcher {
+                    })
+                )[resolvedTemplateSpecificationCID];
+                if (!rawTemplateSpecification)
+                    throw new Error(
+                        `couldn't fetch template specification with cid ${templateSpecificationCID}`
+                    );
+                const parsedTemplateSpecification = JSON.parse(
+                    rawTemplateSpecification
+                );
+                const specification = new TemplateSpecification(
+                    templateSpecificationCID,
+                    parsedTemplateSpecification.name,
+                    parsedTemplateSpecification.description,
+                    parsedTemplateSpecification.tags,
+                    parsedTemplateSpecification.repository,
+                    parsedTemplateSpecification.commitHash
+                );
+                return ResolvedTemplate.from(template, specification);
+            })
+        );
+    }
+
+    protected async resolveEntity<T extends KPIToken | Oracle = KPIToken>({
+        ipfsGatewayURL,
+        entity,
+    }: {
+        entity: T;
+        ipfsGatewayURL: string;
+    }): Promise<T extends KPIToken ? ResolvedKPIToken : ResolvedOracle> {
+        const template = (
+            await this.resolveTemplates({
+                ipfsGatewayURL,
+                templates: [entity.template],
+            })
+        )[0];
+
+        if (!("oracles" in entity))
+            // FIXME: this is ugly but the ts compiler is complaining otherwise (not sure why)
+            return ResolvedOracle.from(entity, template) as T extends KPIToken
+                ? ResolvedKPIToken
+                : ResolvedOracle;
+
+        const oracles = await Promise.all(
+            entity.oracles.map(async (oracle) => {
+                return ResolvedOracle.from(
+                    oracle,
+                    (
+                        await this.resolveTemplates({
+                            ipfsGatewayURL,
+                            templates: [oracle.template],
+                        })
+                    )[0]
+                );
+            })
+        );
+
+        const resolvedKPITokenSpecification = (
+            await this.fetchContentFromIPFS({
+                ipfsGatewayURL,
+                cids: [entity.specificationCID],
+            })
+        )[entity.specificationCID];
+        if (!resolvedKPITokenSpecification)
+            throw new Error(
+                `couldn't fetch kpi token specification with cid ${entity.specificationCID}`
+            );
+
+        return ResolvedKPIToken.from(
+            entity,
+            template,
+            oracles,
+            JSON.parse(resolvedKPITokenSpecification)
+        );
+    }
+
+    public async resolveKPITokens({
+        ipfsGatewayURL,
+        kpiTokens,
+    }: ResolveKPITokensParams): Promise<ResolvedKPITokensMap> {
+        const resolvedKPITokens = await Promise.all(
```

Comment:
Are we sure we want to reject if **any** of the token fetch fails?

Reply:
Should be addressed in c643e9d38303f4dceacedb2dd7f50f8be3aabdda